streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -376,7 +379,7 @@ private void maybeUpdateKeyChangingRepartitionNodeMap() {
         for (final Map.Entry<StreamsGraphNode, Set<StreamsGraphNode>> entry : mergeNodesToKeyChangers.entrySet()) {
             final StreamsGraphNode mergeKey = entry.getKey();
             final Collection<StreamsGraphNode> keyChangingParents = entry.getValue();
-            final Set<OptimizableRepartitionNode> repartitionNodes = new HashSet<>();
+            final Set<OptimizableRepartitionNode> repartitionNodes = new LinkedHashSet<>();
```

Comment:
Keep the `OptimizableRepartitionNode`s in order as they are mapped to merge nodes

Reply:
ditto